pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go

```diff
@@ -106,6 +120,53 @@ func haveOverlap(a1, a2 []string) bool {
 	return false
 }
 
+func (pl *VolumeRestrictions) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod) *framework.Status {
+	return pl.isReadWriteOncePodAccessModeConflict(pod)
+}
+
+// isReadWriteOncePodAccessModeConflict checks if a pod uses a PVC with the ReadWriteOncePod access mode.
+// This access mode restricts volume access to a single pod on a single node. Since only a single pod can
+// use a ReadWriteOncePod PVC, mark any other pods attempting to use this PVC as unschedulable and unresolvable.
+func (pl *VolumeRestrictions) isReadWriteOncePodAccessModeConflict(pod *v1.Pod) *framework.Status {
+	nodeInfos, err := pl.nodeInfoLister.NodeInfos().List()
+	if err != nil {
+		return framework.NewStatus(framework.Error, ErrReasonNodeInfoError)
+	}
+
+	for _, volume := range pod.Spec.Volumes {
+		if volume.PersistentVolumeClaim == nil {
+			continue
+		}
+
+		pvc, err := pl.pvcLister.PersistentVolumeClaims(pod.Namespace).Get(volume.PersistentVolumeClaim.ClaimName)
+		if err != nil {
+			return framework.NewStatus(framework.Error, ErrReasonPVCError)
+		}
+
+		// If ReadWriteOncePod is used, it can be the only one on the PVC since it overrides any
+		// additional access modes. This is also because we do not know what access mode the pod
+		// intends to use if there are multiple.
+		if !(v1helper.ContainsAccessMode(pvc.Spec.AccessModes, v1.ReadWriteOncePod) && len(pvc.Spec.AccessModes) == 1) {
+			continue
+		}
+
+		key := pod.Namespace + "/" + volume.PersistentVolumeClaim.ClaimName
+		for _, nodeInfo := range nodeInfos {
+			refCount, ok := nodeInfo.PVCRefCounts[key]
+			// Enforce ReadWriteOncePod access mode. This is also enforced during volume mount in kubelet.
+			if ok && utilfeature.DefaultFeatureGate.Enabled(features.ReadWriteOncePod) && refCount > 0 {
+				return framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrReasonReadWriteOncePodConflict)
+			}
+		}
+	}
+
+	return framework.NewStatus(framework.Success)
+}
+
+func (pl *VolumeRestrictions) PreFilterExtensions() framework.PreFilterExtensions {
```

Comment:
You actually need to implement this for preemption to work.

This is actually going to be tricky: in preemption, PreFilter doesn't run, for performance reasons. So you have to save some state in `PreFilter`, update it in `AddPod`, `RemovePod` and use that in `Filter`.

There are a million ways to implement that, so you'll have to see what is best performance wise.

Reply:
Resolving, will tackle this in beta